battery_state_broadcaster/include/battery_state_broadcaster/battery_state_broadcaster.hpp

```diff
@@ -0,0 +1,137 @@
+// limitations under the License.
+
+#ifndef BATTERY_STATE_BROADCASTER__BATTERY_STATE_BROADCASTER_HPP_
+#define BATTERY_STATE_BROADCASTER__BATTERY_STATE_BROADCASTER_HPP_
+
+#include <cmath>
+#include <memory>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+#include "controller_interface/controller_interface.hpp"
+#include "rclcpp_lifecycle/node_interfaces/lifecycle_node_interface.hpp"
+#include "rclcpp_lifecycle/state.hpp"
+#include "realtime_tools/realtime_buffer.hpp"
+#include "realtime_tools/realtime_publisher.hpp"
+
+#include <battery_state_broadcaster/battery_state_broadcaster_parameters.hpp>
+#include "control_msgs/msg/battery_states.hpp"
+#include "sensor_msgs/msg/battery_state.hpp"
+
+namespace battery_state_broadcaster
+{
+// // name constants for state interfaces
+// static constexpr size_t STATE_MY_ITFS = 0;
+
+// // name constants for command interfaces
+// static constexpr size_t CMD_MY_ITFS = 0;
+
+/**
+ * \brief Battery State Broadcaster for all or some state in a ros2_control system.
+ *
+ * BatteryStateBroadcaster publishes state interfaces from ros2_control as ROS messages.
+ * The following state interfaces are published:
+ *    <state_joint>/voltage
+ *    <state_joint>/current
+ *    <state_joint>/charge
+ *    <state_joint>/percentage
+ *    <state_joint>/power_supply_status
+ *    <state_joint>/power_supply_health
+ *    <state_joint>/present
+ *
+ * \param state_joints of the batteries to publish.
+ * \param capacity of the batteries to publish.
+ * \param design_capacity of the batteries to publish.
+ * \param power_supply_technology of the batteries to publish.
+ * \param location of the batteries to publish.
+ * \param serial_number of the batteries to publish.
+ *
+ * Publishes to:
+ *
+ * - \b battery_state (sensor_msgs::msg::BatteryState): battery state of the combined battery
+ * joints.
+ * - \b raw_battery_states (battery_state_broadcaster::msg::BatteryStates): battery states of the
+ * individual battery joints.
+ *
+ */
+class BatteryStateBroadcaster : public controller_interface::ControllerInterface
+{
+public:
+  BatteryStateBroadcaster();
+
+  controller_interface::InterfaceConfiguration command_interface_configuration() const override;
+
+  controller_interface::InterfaceConfiguration state_interface_configuration() const override;
+
+  controller_interface::CallbackReturn on_init() override;
+
+  controller_interface::CallbackReturn on_configure(
+    const rclcpp_lifecycle::State & previous_state) override;
+
+  controller_interface::CallbackReturn on_activate(
+    const rclcpp_lifecycle::State & previous_state) override;
+
+  controller_interface::CallbackReturn on_deactivate(
+    const rclcpp_lifecycle::State & previous_state) override;
+
+  controller_interface::return_type update(
+    const rclcpp::Time & time, const rclcpp::Duration & period) override;
+
+  float get_or_nan(int interface_cnt);
+  char get_or_unknown(int interface_cnt);
```

Comment:
please make sure that the return type from these align with where and how you use them. Internally we use doubles for a lot of stuff, using `float` may cause a narrowing conversion.

for char, I see you used `uint8_t` below. If that is what you are looking for, I suggest using that here as return type too. 

Also please double check if the windows compilation (probably on the Ci is enough) is fine with the `uint8_t` type or maybe you need to include stdint.h